atom/browser/api/atom_api_session.cc

```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
 void RunCallbackInUI(const base::Callback<void()>& callback) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI}, callback);
 }
+
 template <typename... T>
 void RunCallbackInUI(const base::Callback<void(T...)>& callback, T... result) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI},
                            base::BindOnce(callback, result...));
 }
 
+void RejectPromise(atom::util::Promise promise, int error) {
+  std::string errmsg =
+      "Failed to size failed with error number: " + std::to_string(error);
+  promise.RejectWithErrorMessage(errmsg);
+}
+
+void RejectPromiseInUI(atom::util::Promise promise, int error) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(RejectPromise, std::move(promise), error));
+}
+
+template <typename T>
+void ResolvePromise(atom::util::Promise promise, T result) {
+  promise.Resolve(result);
+}
+
+template <typename T>
+void ResolvePromiseInUI(atom::util::Promise promise, T result) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(ResolvePromise<T>, std::move(promise), result));
+}
+
 // Callback of HttpCache::GetBackend.
 void OnGetBackend(disk_cache::Backend** backend_ptr,
                   Session::CacheAction action,
-                  const net::CompletionCallback& callback,
+                  const atom::util::CopyablePromise& promise,
                   int result) {
   if (result != net::OK) {
-    RunCallbackInUI(callback, result);
+    RejectPromiseInUI(promise.GetPromise(), result);
```

Comment:
Current api documentation for `getCacheSize` and `clearCache` doesn't return error codes when the operation failed, we can simply follow that and reject with emtpy values. Or if we want to return error codes then better to have more specific error messages, like

```
std::string err_msg = "Failed to retrieve cache backend : " + net::ErrorToString(result);
RejectPromiseInUI(promise.GetPromise(), err_msg);
```